x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/common/chunks/SemanticTextChunkUtils.java

```diff
@@ -0,0 +1,221 @@
+            throw new IllegalStateException(
+                String.format(
+                    Locale.ROOT,
+                    "Invalid content detected for field [%s]: the chunks size is [%d], "
+                        + "but a reference to offset [%d] was found in the result.",
+                    fieldName,
+                    nestedSources.size(),
+                    cand.index()
+                )
+            );
+        }
+        return (String) nestedSources.get(cand.index()).get(SemanticTextField.CHUNKED_TEXT_FIELD);
+    }
+
+    public static List<Query> queries(FieldMapper embeddingsField, TaskType taskType, Query query) throws IOException {
+        return switch (taskType) {
+            case SPARSE_EMBEDDING -> extractSparseVectorQueries(
+                (SparseVectorFieldMapper.SparseVectorFieldType) embeddingsField.fieldType(),
+                query
+            );
+            case TEXT_EMBEDDING -> extractDenseVectorQueries((DenseVectorFieldType) embeddingsField.fieldType(), query);
+            default -> throw new IllegalStateException("Wrong task type for a semantic text field, got [" + taskType.name() + "]");
+        };
+    }
+
+    public static List<OffsetAndScore> extractOffsetAndScores(
+        SearchExecutionContext context,
+        LeafReader reader,
+        SemanticTextFieldMapper.SemanticTextFieldType fieldType,
+        int docId,
+        List<Query> leafQueries
+    ) throws IOException {
+        var bitSet = context.bitsetFilter(fieldType.getChunksField().parentTypeFilter()).getBitSet(reader.getContext());
+        int previousParent = docId > 0 ? bitSet.prevSetBit(docId - 1) : -1;
+
+        BooleanQuery.Builder bq = new BooleanQuery.Builder().add(fieldType.getChunksField().nestedTypeFilter(), BooleanClause.Occur.FILTER);
+        leafQueries.stream().forEach(q -> bq.add(q, BooleanClause.Occur.SHOULD));
+        Weight weight = new IndexSearcher(reader).createWeight(bq.build(), ScoreMode.COMPLETE, 1);
+        Scorer scorer = weight.scorer(reader.getContext());
+        if (scorer == null) {
+            return java.util.List.of();
+        }
+        if (previousParent != -1) {
+            if (scorer.iterator().advance(previousParent) == DocIdSetIterator.NO_MORE_DOCS) {
+                return java.util.List.of();
+            }
+        } else if (scorer.iterator().nextDoc() == DocIdSetIterator.NO_MORE_DOCS) {
+            return List.of();
+        }
+
+        OffsetSourceField.OffsetSourceLoader offsetReader = null;
+        if (fieldType.useLegacyFormat() == false) {
+            var terms = reader.terms(fieldType.getOffsetsField().fullPath());
+            if (terms == null) {
+                // The field is empty
+                return List.of();
+            }
+            offsetReader = OffsetSourceField.loader(terms);
+        }
+
+        List<OffsetAndScore> results = new ArrayList<>();
+        int index = 0;
+        while (scorer.docID() < docId) {
```

Comment:
just to be 100% sure - are we guaranteed that the `scorer.docID()` values will always be in ascending order?